source/win32/low_latency_iocp_context.cpp

```diff
@@ -0,0 +1,682 @@
+    pendingIoQueue_.push_back(op);
+}
+
+void low_latency_iocp_context::release_io_state(vectored_io_state* state) noexcept {
+    assert(is_running_on_io_thread());
+
+    if (state->pendingCompletionNotifications == 0) {
+        // Can be freed immediately.
+
+        if (pendingIoQueue_.empty()) {
+            ioFreeList_.push_front(state);
+        } else {
+            assert(ioFreeList_.empty());
+
+            // Another operation was waiting for an I/O state.
+            // Give the I/O state directly to the operation instead
+            // of adding it to the free-list. This prevents some other
+            // operation from skipping the queue and acquiring it
+            // before this I/O operation can run.
+            auto* op = static_cast<io_operation*>(pendingIoQueue_.pop_front());
+
+            state->parent = op;
+            state->completed = false;
+            state->operationCount = 0;
+            state->pendingCompletionNotifications = 0;
+            op->ioState = state;
+
+            schedule_local(op);
+        }
+    } else {
+        // Mark it to be freed once all of its I/O completion
+        // notifications have been received.
+        state->parent = nullptr;
+    }
+}
+
+void low_latency_iocp_context::schedule_poll_io(io_operation* op) noexcept {
+    assert(is_running_on_io_thread());
+    assert(op != nullptr);
+    assert(op->ioState != nullptr);
+
+    if (op->ioState->pendingCompletionNotifications > 0) {
+        pollQueue_.push_back(op);
+    } else {
+        // No need to poll, schedule straight to the front of the ready-to-run queue.
+        readyQueue_.push_front(op);
+    }
+}
+
+void low_latency_iocp_context::associate_file_handle(handle_t fileHandle) {
+    const HANDLE result = ::CreateIoCompletionPort(
+        fileHandle,
+        iocp_.get(),
+        0, 0);
+    if (result == nullptr) {
+        DWORD errorCode = ::GetLastError();
+        throw std::system_error{
+            static_cast<int>(errorCode), std::system_category(),
+            "CreateIoCompletionPort"};
+    }
+
+    const BOOL ok = ::SetFileCompletionNotificationModes(
+        fileHandle,
+        FILE_SKIP_COMPLETION_PORT_ON_SUCCESS |
+        FILE_SKIP_SET_EVENT_ON_HANDLE);
+    if (!ok) {
+        DWORD errorCode = ::GetLastError();
+        throw std::system_error{
+            static_cast<int>(errorCode), std::system_category(),
+            "SetFileCompletionNotificationModes"};
+    }
+}
+
+void low_latency_iocp_context::io_operation::cancel_io() noexcept {
+    assert(ioState != nullptr);
+
+    // Cancel operations in reverse order so that later operations
+    // are cancelled first and don't accidentally end up with earlier
+    // operations being cancelled and later ones completing due to
+    // a race.
+    for (std::uint16_t i = ioState->operationCount; i != 0; --i) {
+        win32::overlapped* o = &ioState->operations[i-1];
+        (void)::CancelIoEx(fileHandle, reinterpret_cast<OVERLAPPED*>(o));
```

Comment:
Cancelling i/o does not need to occur immediately, it may take time. Yet you are returning immediately. This equals potential memory corruption.

Reply:
This function is called as a request to cancel the operation which is called from the stop_callback in response to a call to `request_stop()`. The caller does not assume that the cancellation will be complete when this function returns.

The storage for the ioState and its OVERLAPPED structures is not released until we have received completion-notifications for all pending I/Os.